mongodb-plugin/src/main/java/io/cdap/plugin/batch/sink/MongoDBBatchSink.java

```diff
@@ -0,0 +1,116 @@
+import com.mongodb.hadoop.MongoOutputFormat;
+import com.mongodb.hadoop.io.BSONWritable;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.data.batch.Output;
+import io.cdap.cdap.api.data.batch.OutputFormatProvider;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.lib.KeyValue;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.batch.BatchSink;
+import io.cdap.cdap.etl.api.batch.BatchSinkContext;
+import io.cdap.plugin.MongoDBConfig;
+import io.cdap.plugin.MongoDBConstants;
+import io.cdap.plugin.common.LineageRecorder;
+import io.cdap.plugin.common.ReferenceBatchSink;
+import io.cdap.plugin.common.ReferencePluginConfig;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link BatchSink} that writes data to MongoDB.
+ * This {@link MongoDBBatchSink} takes a {@link StructuredRecord} in,
+ * converts it to {@link BSONWritable}, and writes it to MongoDB.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name(MongoDBConstants.PLUGIN_NAME)
+@Description("MongoDB Batch Sink converts a StructuredRecord to a BSONWritable and writes it to MongoDB.")
+public class MongoDBBatchSink extends ReferenceBatchSink<StructuredRecord, NullWritable, BSONWritable> {
+
+  private final MongoDBConfig config;
+
+  public MongoDBBatchSink(MongoDBConfig config) {
+    super(new ReferencePluginConfig(config.referenceName));
+    this.config = config;
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws Exception {
+    Configuration conf = new Configuration();
+    String path = conf.get(
+      "mapreduce.task.tmp.dir",
+      conf.get(
+        "mapred.child.tmp",
+        conf.get("hadoop.tmp.dir", System.getProperty("java.io.tmpdir")))) + "/" + UUID.randomUUID().toString();
+
+    if (Objects.nonNull(context.getInputSchema())) {
+      LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+      lineageRecorder.createExternalDataset(context.getInputSchema());
+      List<Schema.Field> fields = context.getInputSchema().getFields();
+      if (fields != null && !fields.isEmpty()) {
+        lineageRecorder.recordWrite("Write", "Wrote to MongoDB collection.",
+                                    fields.stream().map(Schema.Field::getName).collect(Collectors.toList()));
+      }
+    }
+
+    context.addOutput(Output.of(config.referenceName, new MongoDBOutputFormatProvider(config, path)));
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, BSONWritable>> emitter)
+    throws Exception {
+    BasicDBObjectBuilder bsonBuilder = BasicDBObjectBuilder.start();
```

Comment:
this transformation should also be excluded to separate class, usually with Transformer suffix

Reply:
[RecordToBSONWritableTransformer](https://github.com/vladhlinskiy/database-plugins/pull/14/files#diff-fb703efbf417ef98f9f95c778e0b0125R30) class created.